pageserver/src/task_mgr.rs

```diff
@@ -0,0 +1,457 @@
+    // events from etcd, decides which safekeeper to connect to. It spawns a
+    // separate WalReceiverConnection task to handle each connection.
+    WalReceiverManager,
+
+    // Handles a connection to a safekeeper, to stream WAL to a timeline.
+    WalReceiverConnection,
+
+    // Garbage collection worker. One per tenant
+    GarbageCollector,
+
+    // Compaction. One per tenant.
+    Compaction,
+
+    // Task that flushes frozen in-memory layers to disk
+    LayerFlushTask,
+
+    // Task that manages the remote upload queue
+    StorageSync,
+
+    // task that handles the initial downloading of all tenants
+    InitialLoad,
+
+    // task that handles attaching a tenant
+    Attach,
+}
+
+#[derive(Default)]
+struct MutableTaskState {
+    /// Tenant and timeline that this task is associated with.
+    tenant_id: Option<ZTenantId>,
+    timeline_id: Option<ZTimelineId>,
+
+    /// Handle for waiting for the task to exit. It can be None, if the
+    /// the task has already exited.
+    join_handle: Option<JoinHandle<()>>,
+}
+
+struct PageServerTask {
+    #[allow(dead_code)] // unused currently
+    task_id: PageserverTaskId,
+
+    kind: TaskKind,
+
+    name: String,
+
+    // To request task shutdown, send 'true' to the channel to notify the task.
+    shutdown_tx: watch::Sender<bool>,
+
+    mutable: Mutex<MutableTaskState>,
+}
+
+/// Launch a new task
+/// Note: if shutdown_process_on_error is set to true failure
+///   of the task will lead to shutdown of entire process
+pub fn spawn<F>(
+    runtime: &tokio::runtime::Handle,
+    kind: TaskKind,
+    tenant_id: Option<ZTenantId>,
+    timeline_id: Option<ZTimelineId>,
+    name: &str,
+    shutdown_process_on_error: bool,
+    future: F,
+) -> PageserverTaskId
+where
+    F: Future<Output = anyhow::Result<()>> + Send + 'static,
+{
+    let (shutdown_tx, shutdown_rx) = watch::channel(false);
+    let task_id = NEXT_TASK_ID.fetch_add(1, Ordering::Relaxed);
+    let task = Arc::new(PageServerTask {
+        task_id: PageserverTaskId(task_id),
+        kind,
+        name: name.to_string(),
+        shutdown_tx,
+        mutable: Mutex::new(MutableTaskState {
+            tenant_id,
+            timeline_id,
+            join_handle: None,
+        }),
+    });
+
+    TASKS.lock().unwrap().insert(task_id, Arc::clone(&task));
+
+    let mut task_mut = task.mutable.lock().unwrap();
+
+    let task_name = name.to_string();
+    let task_cloned = Arc::clone(&task);
+    let join_handle = runtime.spawn(task_wrapper(
+        task_name,
+        task_id,
+        task_cloned,
+        shutdown_rx,
+        shutdown_process_on_error,
+        future,
+    ));
+    task_mut.join_handle = Some(join_handle);
+    drop(task_mut);
+
+    // The task is now running. Nothing more to do here
+    PageserverTaskId(task_id)
+}
+
+/// This wrapper function runs in a newly-spawned task. It initializes the
+/// task-local variables and calls the payload function.
+async fn task_wrapper<F>(
```

Comment:
I think, this and `TaskHandle` are the same thing and eventually should be merged.

After that when we have all tasks under `task_mgr` control and enough data in them, I would imagine that `SHUTDOWN_RX` would not be needed to be public at all: there should be some global `select!` per task that checks for cancellations and allows the tasks to be run to the end (or panic), checking that and rescheduling the whole task if needed.

So, general task orchestration should be concentrated in one place, invisibly to the tasks themselves.

But sure that's the future, good that we're able to recreate the present already.

Reply:
Come to think of it, cannot we already use a `select!` in this particular function?

Reply:
I don't want cancellation to happen implicitly at any `await` call, I'd prefer it to happen only at explicitly chosen points.

Reply:
> I think, this and TaskHandle are the same thing and eventually should be merged.

Yeah, there's a lot of overlap. I tried to remove TaskHandle when I started on this, but it was not so simple. The TaskHandle provides the channel between the two tasks, and allows you to wait on the particular Task to finish.  The code in `task_mgr.rs` doesn't allow you to do that currently.

It would make sense to do some more refactoring here, but I'm not sure what exactly.

Reply:
> I don't want cancellation to happen implicitly at any await call, I'd prefer it to happen only at explicitly chosen points.

Can't give this PR much of well-deserved attention just yet, but this idea is worth commenting on.

Due to the implementation of async in rust, you can only hope things will work out the way you want. There's nothing to prevent any intermediate async combinator (be it `tokio::select!` or anything else) from cancelling your future at an arbitrary point of time. If it doesn't call `poll()` on your future, it won't move any further; and thus your future might be destroyed before getting over an `.await`.

This is exactly why it's very important to release any resources in destructors or wrappers like `scopeguard::defer!`, because all we know is that destructors will most likely be executed (but even that is not guaranteed, e.g. `std::mem::forget`, rc loops and such). Unfortunately, there's no such thing as "async drop" just yet, which is another missing piece in the async story.

This is to say that we should probably plan accordingly and try writing code that can survive this. Otherwise it's a matter of time before we find some really nasty bugs due to being unable to uphold this invariant.

Reply:
The current best practice is to document any cancellation-safety specific behavior. For example see [this](https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety) tokio doc on `select!`. It explicitly mentions some of the cancellation-safe/unsafe functions. And other functions typically mention cancellation safety in their doc comment.  We can do the same thing. Also there are some patterns (and libraries) we can use to help us go around these problems. See the [blog post](http://smallcultfollowing.com/babysteps/blog/2022/06/13/async-cancellation-a-case-study-of-pub-sub-in-mini-redis/) with in-depth explanation. Also see [this blog post](https://boats.gitlab.io/blog/post/poll-drop/) describing problems which arise on the way of introducing async destructors.

Reply:
Agreed that we should try to be cancellation-safe. As in, if a task gets cancelled unexpectedly, it should not lead to data corruption, for example. Still, I think it's better to avoid cancellation at arbitrary points. Unexpected cancellation could easily lead to unexpected errors or other weird things.

The flip-side is that we might miss checking for shutdown in some long-running codepath, so that shutdown takes longer than you would expect.